linux: fix invalid cross-device link error

sjet47 opened the pull request:
This PR fix the "invalid cross-device link" error occurred in linux when trying to write the settings file atomically, like when click the "Enable vim mode" checkbox at first start.

```plain
[2024-02-26T22:59:25+08:00 ERROR util] .../zed/crates/settings/src/settings_file.rs:135: Failed to write settings to file "/home/$USER/.config/zed/settings.json"

Caused by:
0: failed to persist temporary file: Invalid cross-device link (os error 18)
1: Invalid cross-device link (os error 18)
```

Currently the `fs::RealFs::atomic_write()` method write to a temp file created with `NamedTempFile::new()` and then call `persist()` method to write to the config file path, which actually do a `rename` syscall under the hood. As the [issue](https://github.com/Stebalien/tempfile/issues/245) said

> `NamedTempFile::new()` will create a temporary file in your system's temporary file directory. You need `NamedTempFile::new_in()`.

The temporary file directory in linux is in `/tmp`, which is mounted to `tmpfs` filesystem, and in most case(all case I guess) `$HOME/.config/zed` is mounted to a different filesystem. And the `rename` syscall between different filesystems will return a `EXDEV` errno, as described in the man page [rename(2)](https://man7.org/linux/man-pages/man2/renameat2.2.html):

```plain
       EXDEV  oldpath and newpath are not on the same mounted
              filesystem.  (Linux permits a filesystem to be mounted at
              multiple points, but rename() does not work across
              different mount points, even if the same filesystem is
              mounted on both.)
```

And as the issue above said, use a different temp dir with `NamedTempFile::new_in()` for linux platform might be a solution, since the `rename` syscall provides atomicity.

Release Notes:
 - Fix `settings.json` save failed with invalid cross-device link error in linux

sjet47 committed:
linux: fix invalid cross-device link error

ASjet commented:
@cla-bot check

sjet47 committed:
docs: add comment for change

romgrk commented on crates/util/src/paths.rs:
```diff
@@ -44,6 +44,7 @@ lazy_static::lazy_static! {
     pub static ref LOG: PathBuf = LOGS_DIR.join("Zed.log");
     pub static ref OLD_LOG: PathBuf = LOGS_DIR.join("Zed.log.old");
     pub static ref LOCAL_SETTINGS_RELATIVE_PATH: &'static Path = Path::new(".zed/settings.json");
+    pub static ref TEMP_DIR: PathBuf = CONFIG_DIR.join("tmp");
```
Small nit, but the `XDG_CACHE_DIR` is a better candidate for this.

https://wiki.archlinux.org/title/XDG_Base_Directory

romgrk replied:
Other alternative could be to write the file to the same directory as the target path, it would ensure that it's on the same filesystem.

ASjet replied:
Good point! write to the same directory would be a better choice

sjet47 committed:
use current dir as temp dir (#8437)

In atomic_write use the write path as the temp dir can ensure the temp file is in the same filesystem. Use XDG_CACHE_DIR as a fallback

sjet47 committed:
docs: update comment (#8437)

mikayla-maki reviewed (approved):
Thank you!

mikayla-maki commented on crates/zed/src/main.rs:
```diff
@@ -481,6 +481,11 @@ fn init_paths() {
     std::fs::create_dir_all(&*util::paths::LANGUAGES_DIR).expect("could not create languages path");
     std::fs::create_dir_all(&*util::paths::DB_DIR).expect("could not create database path");
     std::fs::create_dir_all(&*util::paths::LOGS_DIR).expect("could not create logs path");
+
```
Could we remove this comment?

ASjet replied:
Sure!

mikayla-maki reviewed (changes requested):
Thank you!

mikayla-maki commented:
It also looks like this needs a cargo fmt :)

sjet47 committed:
style: remove unnecessary comment and do cargo fmt (#8437)
